File: Device_control/PI-E709/PI_codes/PI-Software-Suite-C-990.CD1/Development/C#/Samples/General Controller/TwoAxesSample/TwoAxesSample.cs
/** <copyright file="TwoAxisSample.cs" company="Physik Instrumente (PI) SE & Co. KG">
 *  (c)2015-2020 Physik Instrumente (PI) SE & Co. KG
 *  Software products that are provided by PI are subject to the General Software License Agreement of
 *  Physik Instrumente (PI) SE & Co. KG and may incorporate and/or make use of third-party software components.
 *  For more information, please read the General Software License Agreement and the Third Party Software Note linked below.
 *
 *  General Software License Agreement :
 *  http://www.physikinstrumente.com/download/EULA_PhysikInstrumenteGmbH_Co_KG.pdf
 *  Third Party Software Note :
 *  http://www.physikinstrumente.com/download/TPSWNote_PhysikInstrumenteGmbH_Co_KG.pdf
 *  </copyright>*/

/**<remarks>
 * !! MOTION WARNING !!
 *
 * This sample connects to a controller with at least 2 axes
 * and demonstrate how to use 2 axis.
 *
 * Command a absolute movement with 2 axes with one command and afterwards
 * demonstrate a movement with 2 axes sequentially.
 *
 * Both axes have to be initialized and referenced.
 * </remarks> */

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

using PI;

namespace TwoAxisSample
{
    public enum ConnectionType
    {
        Dialog,
        Rs232,
        Tcpip,
        Usb
    }

    internal class GcsCommandError : Exception
    {
        public GcsCommandError( string message )
            : base(message)
        {
        }
    }

    /** <summary>
    *  This class runs the sample for a controller with at least 2 axis, i.e. C-884, C-887.
    *  Call the Run-method to execute this sample.
    *  Please choose the right configuration for your setup, there is one common configuration
    *  working with most of PI controllers and one configuration for the Hexapod controller C-887
    * </summary>*/
    public class TwoAxisSample
    {
        /// <remarks>
        /// select the wave generator configuration here: 'Common' or 'C-887'
        /// </remarks>
        #region controller specific configuration
        public static Dictionary<string, string> AxisIdentifier { get; }
            = new Dictionary<string, string>
            {
                {"Common", "1 2"},
                {"C-887", "X Y"},
            };

        private static readonly string Axis = AxisIdentifier["Common"];
        private const int PI_NUMBER_OF_AXIS = 2;
        #endregion

        #region fields
        internal int ControllerId = -1;
        private const int PI_RESULT_FAILURE = 0;
        private const int PI_TRUE = 1;
        private const int PI_FALSE = 0;
        #endregion

        #region public methods
        public void Run()
        {
            try
            {
                OpenConnection(ConnectionType.Dialog);
                PrintControllerIdentification();

                var state = Enumerable.Repeat(PI_TRUE, PI_NUMBER_OF_AXIS).ToArray();
                SetServoState(Axis, state);

                double[] minPosLimits = GetMinPositionLimit(Axis);
                double[] maxPosLimits = GetMaxPositionLimit(Axis);
                var travelRange = CalculateTravelRange(minPosLimits, maxPosLimits);

                MoveAbsoluteToNearMinimum(minPosLimits, travelRange);
                WaitForMotionDone(Axis).Wait(8000);
                RelativeMoves( maxPosLimits, travelRange );

                state = Enumerable.Repeat(PI_FALSE, PI_NUMBER_OF_AXIS).ToArray();
                SetServoState(Axis, state);
            }
            catch (GcsCommandError e)
            {
                PrintErrorMessage(e);
            }
            finally
            {
                DisconnectController();
            }
        }
        #endregion

        #region private methods
        private void MoveAbsoluteToNearMinimum(double[] minPosLimits, IEnumerable<double> travelRange)
        {
            if (travelRange == null) throw new ArgumentNullException(nameof(travelRange));
            if (minPosLimits == null) throw new ArgumentNullException(nameof(minPosLimits));
            var targetPos = CalculateTargetPosition(minPosLimits, travelRange, 1.0);
            MoveToAbsoluteTarget(targetPos);
        }

        private void RelativeMoves(double[] maxPosLimits, IEnumerable<double> travelRange)
        {
            var aAxis = Axis.Split(' ').ToArray();
            var aTravelRange = travelRange.ToArray();
            var distances = Array.ConvertAll(aTravelRange, x => x * 0.1);

            while (true)
            {
                for (var idx = 0; idx < aAxis.Length; idx++)
                {
                    Console.WriteLine($"Moving axis {aAxis[idx]} for {distances[idx]} units");

                    try
                    {
                        RelativeMove(aAxis[idx], new[] { distances[idx] });
                    }
                    catch (GcsCommandError e)
                    {
                        // we expect this to fail, since the distances are increased every time
                        Console.WriteLine($"MVR failed.Current axis: {aAxis[idx]}; Value: {distances[idx]}");
                        PrintErrorMessage(e);
                        return;
                    }

                    WaitForMotionDone(aAxis[idx]).Wait(8000);
                }

                distances = Array.ConvertAll(distances, x => x * -1.0 * 1.2);
            }
        }

        private double[] CalculateTravelRange(double[] minPosLimits, double[] maxPosLimits)
        {
            var travelRange = (
                    from position
                        in minPosLimits.Zip(maxPosLimits, (min, max) => new {Min = min, Max = max})
                    let result = Math.Abs(position.Max) + Math.Abs(position.Min)
                    select result
                )
                .ToArray();
            return travelRange;
        }

        private void MoveToAbsoluteTarget(double[] targetPos)
        {
            if (GCS2.MOV(ControllerId, Axis, targetPos) == PI_RESULT_FAILURE)
            {
                throw new GcsCommandError("Unable to move to target position.");
            }
        }

        private void RelativeMove(string axis, double[] targetPos)
        {
            if (GCS2.MVR(ControllerId, axis, targetPos) == PI_RESULT_FAILURE)
            {
                throw new GcsCommandError("Unable to move to target position.");
            }
        }

        private async Task WaitForMotionDone ( string axis )
        {
            var isMoving = Enumerable.Repeat(1, axis.Split(' ').Length ).ToArray();

            while ( Array.Exists(isMoving, element => element != 0 ))
            {
                if (GCS2.IsMoving(ControllerId, axis, isMoving ) == PI_RESULT_FAILURE )
                {
                    throw new GcsCommandError("Unable to query movement status.");
                }

                await Task.Delay(10);
            }

        }

        private static double[] CalculateTargetPosition(IEnumerable<double> minPosLimits
            , IEnumerable<double> travelRange
            , double modifier )
        {
            var targetPos = (
                    from position
                        in minPosLimits.Zip(travelRange, (limit, range) => new {Limit = limit, Range = range})
                    let result = position.Limit + (position.Range * modifier * 0.1)
                    select result
                )
                .ToArray();
            return targetPos;
        }

        private double[] GetMaxPositionLimit(string axis)
        {
            var maxPosLimits = new double[PI_NUMBER_OF_AXIS];

            if (GCS2.qTMX (ControllerId, axis, maxPosLimits) == PI_RESULT_FAILURE )
            {
                throw new GcsCommandError("Unable to get maximum position limit.");
            }
            return maxPosLimits;
        }

        private double[] GetMinPositionLimit(string axis)
        {
            var minPosLimits = new double[PI_NUMBER_OF_AXIS];

            if (GCS2.qTMN (ControllerId, axis, minPosLimits) == PI_RESULT_FAILURE )
            {
                throw new GcsCommandError("Unable to get minimum position limit.");
            }

            return minPosLimits;
        }

        private void SetServoState(string axis, int[] state)
        {
            if (GCS2.SVO(ControllerId, axis, state) == PI_RESULT_FAILURE)
            {
                throw (new GcsCommandError("SVO failed. Exiting"));
            }
        }

        private void PrintControllerIdentification()
        {
            var controllerIdentification = new StringBuilder(1024);

            if (GCS2.qIDN(ControllerId, controllerIdentification, controllerIdentification.Capacity) ==
                PI_RESULT_FAILURE)
            {
                throw (new GcsCommandError("qIDN failed. Exiting."));
            }

            Console.WriteLine("qIDN returned: " + controllerIdentification);
        }

        private void OpenConnection(ConnectionType type, string hostName = "localhost")
        {
            switch (type)
            {
                case ConnectionType.Dialog:
                    ControllerId = GCS2.InterfaceSetupDlg("");
                    break;
                case ConnectionType.Rs232:
                    ControllerId = GCS2.ConnectRS232(1, 115200);
                    break;
                case ConnectionType.Tcpip:
                    ControllerId = GCS2.ConnectTCPIP(hostName, 50000);
                    break;
                case ConnectionType.Usb:
                {
                    var connectedUsbController = new StringBuilder(1024);
                    var noDevices = GCS2.EnumerateUSB(connectedUsbController, 1024, "");

                    Console.WriteLine("Found: " + noDevices + " USB controllers, connecting to the first one:");
                    Console.Write(connectedUsbController + "\n");

                    ControllerId = GCS2.ConnectUSB(
                        connectedUsbController
                            .ToString()
                            .Split(new[] {Environment.NewLine}, StringSplitOptions.None)
                            [0]);
                    break;
                }
                default:
                    throw new ArgumentOutOfRangeException(nameof(type), type, null);
            }

            if (0 <= ControllerId) return;

            var connType = type.Equals(ConnectionType.Dialog) ? type.ToString() : type.ToString().ToUpper();
            throw new GcsCommandError("Connection to " + connType + " failed. Exiting.");
        }

        private void DisconnectController()
        {
            if (ControllerId >= 0)
            {
                GCS2.CloseConnection(ControllerId);
                ControllerId = -1;
            }

            Console.WriteLine("Close connection.");
        }

        private void PrintErrorMessage(GcsCommandError e)
        {
            var errorCode = GCS2.GetError(ControllerId);
            var errorMessage = new StringBuilder(1024);
            GCS2.TranslateError(errorCode, errorMessage, errorMessage.Capacity);
            Console.WriteLine("ERROR: " + e.Message + "\n - Reported error code: " + errorCode + ": \n\t" +
                              errorMessage);
        }
        #endregion
    }
}

File: Device_control/PI-E709/PI_codes/PI-Software-Suite-C-990.CD1/Development/C++/Samples/General Controller/TwoAxesSample/TwoAxesSample.cpp
/**
    (c)2015 Physik Instrumente (PI) SE & Co. KG
    Software products that are provided by PI are subject to the General Software License Agreement of 
    Physik Instrumente (PI) SE & Co. KG and may incorporate and/or make use of third-party software components. 
    For more information, please read the General Software License Agreement and the Third Party Software Note linked below.

    General Software License Agreement :
    http://www.physikinstrumente.com/download/EULA_PhysikInstrumenteGmbH_Co_KG.pdf
    Third Party Software Note :
    http://www.physikinstrumente.com/download/TPSWNote_PhysikInstrumenteGmbH_Co_KG.pdf
**/

/**
    !! MOTION WARNING !!

    This sample connects to a two-axis controller.
    Afterwards, it moves the two axes forward and backward repeatedly.
    The two axes have to be referenced before starting this sample.
	Note: The example does not check whether the target position is within the permissible working space of the positioner.
    
**/

#include "TwoAxesSample.h"

#include <algorithm>

int main ()
{
    int iD = OpenConnection (Dialog);

    if (iD < 0)
    {
        std::cout << "Unable to connect.";
        return 1;
    }

    try
    {
        PrintControllerIdentification (iD);

        // based on your controller type, use the appropriate constant
        std::string sAxis = commonAxesIdentifier;
        
        SetServoState (iD, sAxis, SERVO_ON);
        std::vector<double> minPosLimits = GetMinPositionLimit(iD, sAxis);
        std::vector<double> maxPosLimits = GetMaxPositionLimit(iD, sAxis);
        std::vector<double> travelRange = CalculateTravelRange(minPosLimits, maxPosLimits);

        MoveAbsoluteToNearMinimum (iD, sAxis, minPosLimits, travelRange);
		WaitForMotionDone(iD, "");
        RelativeMoves (iD, sAxis, maxPosLimits, travelRange);
        SetServoState (iD, sAxis, SERVO_OFF);

        PI_CloseConnection (iD);
        std::cout << "Closing connection." << std::endl;
    }
    catch (std::runtime_error e)
    {
        CloseConnectionWithComment (iD, e.what ());
        return 1;
    }

    return 0;
}

int OpenConnection (ConnectionType type, std::string hostName)
{
    int iD = 0;
    std::string connType;

    switch (type)
    {
        case Dialog:
            iD = PI_InterfaceSetupDlg ("");
            connType = "Dialog";
            break;
        case RS232:
            iD = PI_ConnectRS232 (1, 115200);
            connType = "RS232";
            break;
        case TCPIP:
            iD = PI_ConnectTCPIP (hostName.c_str (), 50000);
            connType = "TCPIP";
            break;
        case USB:
            char connectedUsbController[1024];
            int noDevices = PI_EnumerateUSB (connectedUsbController, 1024, "");
            std::cout << "Found " << noDevices << " controllers connected via USB, connecting to the first one:" << std::endl << connectedUsbController << std::endl;
            char* pos = strchr (connectedUsbController, '\n');

            if (pos)
            {
                *pos = '\0';
            }

            iD = PI_ConnectUSB (connectedUsbController);
            connType = "USB";
            break;
    }

    if (iD < 0)
    {
        ReportError (iD);
        std::cout << "Connect " << connType << ": ERROR" << std::endl;
    }

    return iD;
}

void PrintControllerIdentification (int iD)
{
    char szIDN[200];

    if (!PI_qIDN (iD, szIDN, 199))
    {
        std::runtime_error ("qIDN failed. Exiting.");
    }
    std::cout << "qIDN returned: " << szIDN << std::endl;
}

void CloseConnectionWithComment (int iD, std::string comment)
{
    std::cout << comment << std::endl;

    ReportError (iD);
    PI_CloseConnection (iD);
}

std::vector<double> GetCurrentPosition(int ID, std::string axes)
{
    double currentPositions[NUMBER_OF_AXES];

    if (!PI_qPOS (ID, axes.c_str (), currentPositions))
    {
        throw std::runtime_error ("Unable to query current position.");
    }

    return std::vector<double> (currentPositions, currentPositions + NUMBER_OF_AXES);
}

std::vector<double> GetMinPositionLimit(int ID, std::string axis)
{
    double minPosLimit[NUMBER_OF_AXES];

    if (!PI_qTMN(ID, axis.c_str(), minPosLimit))
    {
        throw std::runtime_error("Unable to get minimum position limit.");
    }

    return std::vector<double> (minPosLimit, minPosLimit + NUMBER_OF_AXES);
}

std::vector<double> GetMaxPositionLimit(int ID, std::string axis)
{
    double maxPosLimit[NUMBER_OF_AXES];

    if (!PI_qTMX(ID, axis.c_str(), maxPosLimit))
    {
        throw std::runtime_error("Unable to get maximum position limit.");
    }

    return std::vector<double> (maxPosLimit, maxPosLimit + NUMBER_OF_AXES);

}

std::vector<double> CalculateTravelRange(std::vector<double> minPosLimits, std::vector<double> maxPosLimits)
{
    double travelRange[NUMBER_OF_AXES];

    for (int idx = 0; idx < NUMBER_OF_AXES; idx++)
    {
        travelRange[idx] = std::abs (maxPosLimits.at(idx)) + std::abs (minPosLimits.at(idx));
    }

    return std::vector<double> (travelRange, travelRange + NUMBER_OF_AXES);
}

bool IsNotNearLimit(std::string axes, std::vector<double> maxPositionLimits, std::vector<double> currentPositions)
{
    bool isNotNearLimit = true;

    for (int idx = 0; idx < NUMBER_OF_AXES; idx++)
    {
        isNotNearLimit = isNotNearLimit && maxPositionLimits.at(idx) * 0.75 > currentPositions.at(idx);
    }

    return isNotNearLimit;
}

void MoveAbsoluteToNearMinimum(int ID, std::string axis, std::vector<double> minPosLimits, std::vector<double> travelRange)
{
    double targetPositions[NUMBER_OF_AXES];

    for (int idx = 0; idx < NUMBER_OF_AXES; idx++)
    {
        targetPositions[idx] = minPosLimits.at(idx) + (travelRange.at(idx) * 0.1);
    }

    if (!PI_MOV (ID, axis.c_str (), targetPositions))
    {
        throw std::runtime_error ("Unable to move axes near minimum position.");
    }
}

void RelativeMoves (int ID, std::string axis, std::vector<double> maxPosLimits, std::vector<double> travelRange)
{
	std::vector<std::string> vAxis;
	vAxis.push_back(axis.substr(0, 1));
	vAxis.push_back(axis.substr(axis.size() - 1, 1));

	std::vector<double> distances = travelRange;
	std::transform(distances.begin(), distances.end(), distances.begin(), [](double x) {return x * 0.1; });
	while (true)
	{
		for (size_t idx = 0; idx < vAxis.size(); idx++)
		{
			std::cout << "Moving axis " << vAxis[idx] << " for " << distances[idx] << " units" << std::endl;

			try
			{
				RelativeMove(ID, vAxis[idx], distances[idx]);
			}
			catch (...)
			{
				// we expect this to fail, since the distances are increased every time
			    std::string msg ("MVR failed. Current axis: ");
			    msg += vAxis[idx].c_str();
			    msg += "; Value: ";
			    msg += std::to_string ( (long long)distances[idx]);
				std::cout << msg << std::endl;
				return;
			}

			WaitForMotionDone(ID, vAxis[idx]);
			Sleep(30);
		}
		std::transform(distances.begin(), distances.end(), distances.begin(),
			[&](double x) {return x * -1.0 * 1.2; });
	}
}

void RelativeMove(int ID, std::string axis, double offset)
{
	if (!PI_MVR (ID, axis.c_str(), &offset))
	{
		throw std::runtime_error("Cannot move to target.");
	}
}

bool WaitForMotionDone (int ID, std::string axis)
{
    BOOL isMoving = TRUE;

    while (isMoving)
    {
        if (!PI_IsMoving (ID, axis.c_str (), &isMoving))
        {
            throw std::runtime_error ("Unable to retrieve motion state.");
        }

        Sleep (100);
    }

    return true;
}

void SetServoState (int ID, std::string axis, BOOL state)
{
    if (!PI_SVO (ID, axis.c_str (), &state))
    {
        throw std::runtime_error ("SVO failed. Exiting");
    }

    std::cout << "Turned servo " << (state ? "on" : "off") << std::endl;
}

void ReportError (int iD)
{
    int err = PI_GetError (iD);
    char szErrMsg[300];

    if (PI_TranslateError (err, szErrMsg, 299))
    {
        std::cout << "Error " << err << " occurred: " << szErrMsg << std::endl;
    }
}
